### PR TITLE
add commands for opening "find references" in bottom or quick panel

### PR DESCRIPTION
**Pull request by rchl**
Include an additional variant of the `Find references` command in Command Palette and menus that allows to show references in the way that is opposite to the default behavior (which depends on `show_references_in_quick_panel`).

So when `show_references_in_quick_panel` is enabled, we'll show `Find References (in bottom panel)` in addition to the default `Find References` and if `show_references_in_quick_panel` is disabled we'll show `Find References (in quick panel)` so that both variants are always accessible.

The reason for that is that while showing references in the quick panel is typically the preferred way, one might want to occasionally show those in the bottom panel to be able to keep the list as a reference when wanting to update all results manually.

### CODE REVIEW

**Comment by jwortmann** [comment]
For the context menu, would it be better to change to the alternative display method if a modifier key is pressed when you select the "Find References" item? I think the additional entry is okay for the command palette, but I think it's not the best UI to have two of the items with basically the same functionality in the context menu. Modifier keys for the context menu should work fine (you can confirm by adding `print(event)` in `run()`): 
```
{'modifier_keys': {}, 'x': 654.5, 'y': 518.5, 'text_point': 2176}
{'modifier_keys': {'primary': True, 'ctrl': True}, 'x': 652.5, 'y': 518.5, 'text_point': 2176}
```

**Comment by rchl** [comment]
That would make the feature pretty much impossible to discover though...

**Comment by jwortmann** [comment]
I gues most users are satisfied with having the global "how_references_in_quick_panel": true/false setting. The only request for that in the issue tracker I can see is from you: https://github.com/sublimelsp/LSP/issues/2084
I don't want to say that it isn't a valid request, but we could mention it in a sentence in the docs and if someone searches or asks for it, it shouldn't be too hard to find out about it. There are also no additional context menu items for the other command arguments; for example we have the key binding for side-by-side (using the default keys from ST) even commented out, and we have no explicit context menu item "Find References (side by side)"
https://github.com/sublimelsp/LSP/blob/8023408dfd7917d23f89c9bb3f7356a4d2d1d895/Default.sublime-keymap#L70C1-L76C10

**Comment by jwortmann** [comment]
Actually we could even use the modifier key too for the side-by-side mode. This would only be relevant if it is displayed in the quick panel. I.e. activate side-by-side if <kbd>Shift</kbd> is pressed (like in the default key binding) and change panel mode (bottom/quick panel) if <kbd>Primary</kbd> is pressed.

File: plugin/references.py [hunk 1]
```diff
@@ -44,10 +48,18 @@ def is_visible(
...
+        show_in: Optional[ShowInArgument] = None,
     ) -> bool:
+        # We include "in bottom panel" and "in quick panel" variants of `LSP: Find References` in the Command Palette
+        # but we only show the one that is not the same as the default one (per `show_references_in_quick_panel`
+        # setting).
+        if show_in == 'bottom_panel' and not userprefs().show_references_in_quick_panel or \
+                show_in == 'quick_panel' and userprefs().show_references_in_quick_panel:
+            return False
         if self.applies_to_context_menu(event):
-            return self.is_enabled(event, point, side_by_side, fallback)
```
Comment (rchl) [inline plugin/references.py:50]: BTW. Here we had a bug - passed "fallback" to "force_group".

**Comment by rchl** [comment]
> activate side-by-side if Shift is pressed (like in the default key binding)

Actually, the `primary` key is used to trigger side-by-side. Which does give us a lot of wiggle room.

That means that ideally for the side-by-side triggered from the context menu we'd also need to use primary (ctrl on win, cmd on mac) but then what would we use for "default-inverted" action? Shift? Doesn't feel quite right.
- "alternative" action:

**Comment by rchl** [comment]
I've added support for holding `primary` key while triggering context menu item but haven't added support for triggering "non-default" behavior. I guess the only option is to go with `shift` here?

**Comment by jwortmann** [comment]
Yes, I confused the <kbd>Shift</kbd> and <kbd>Primary</kbd> keys above. I think it is better now than having additional menu items.

A few more suggestions for the labels and names:
- Is there maybe a better name for the "show_in" argument? I'd prefer finding one or two words to describe the argument (usually as a noun) if possible, instead of looking like an incomplete sentence. For example
  - "quick_panel": true/false (this would be bad if another mode should be added in the future though)
  - "display_mode"
  - "presentation"
  - or maybe you have another idea with a better name
- I would probably use "output_panel" instead of "bottom_panel" for the option and for the command palette label. Because it may not be restricted to the bottom forever, if ST adds more flexibility for the panels in a future release (https://github.com/sublimehq/sublime_text/issues/1451)
- I think the word "in" should be removed from the command palette labels, currently it reads a bit as the command would do the search inside the panel. Or alternatively add "show in"
  - "LSP: Find References (Quick Panel)"
  - "LSP: Find References (Output Panel)"

**Comment by rchl** [comment]
The reason I've used "bottom panel" is to match the description of the setting:

```
  // Show symbol references in Sublime's quick panel instead of the bottom panel.
  "show_references_in_quick_panel": true,
```

but I'm fine with changing both.

**Comment by rchl** [comment]
> * "quick_panel": true/false

Don't like this since it's not clear what "false" means.
I don't know if I have good suggestions... `output` or `target` come to mind

**Comment by predragnikolic** [comment]
I tried pressing `shift+enter` in the context menu, but it does nothing for me.
Only when I press just `enter`, then it works (end of the gif when the quick panel opens) 

output

I did a little search in my keybindings,  and found out that I had `shift+enter` in it, so I commented that out, but it still doesn't work.
```
    // {
    //     "command": "lsp_commit_completion_with_opposite_insert_mode",
    //     "keys": [
    //         "shift+enter"
    //     ],
    //     "context": [
    //         {
    //             "key": "auto_complete_visible",
    //             "operator": "equal",
    //             "operand": true
    //         },
    //         {
    //             "key": "lsp.session_with_capability",
    //             "operator": "equal",
    //             "operand": "completionProvider"
    //         }
    //     ]
    // },
```

**Comment by rchl** [comment]
Do other menu items work? If it doesn't work with those that don't handle shift specifically (so basically most of them) then I would assume it's a Linux/Window Manager thing.

File: plugin/references.py [hunk 2]
```diff
@@ -113,17 +130,24 @@ def _handle_response(
...
-            else:
-                self._show_references_in_output_panel(word, session, response)
-        else:
+        if not response:
             self._handle_no_results(fallback, side_by_side)
+            return
+        modifier_keys = (event or {}).get('modifier_keys', {})
+        show_in_quick_panel = show_in == 'quick_panel' or show_in is None and userprefs().show_references_in_quick_panel
+        if modifier_keys.get('shift'):
+            show_in_quick_panel = not show_in_quick_panel
```
Comment (rchl) [inline plugin/references.py:144]: I've realized that I should only be handling shift when "show_in" is not specified.

**Comment by rchl** [comment]
> > * "quick_panel": true/false
> 
> Don't like this since it's not clear what "false" means. I don't know if I have good suggestions... `output` or `target` come to mind

`show_mode`?

**Comment by jwortmann** [comment]
> `show_mode`?

Hm... this sounds a bit like a "TV show" or something. I like your "output" suggestion, or "output_mode", or "display_mode".

The command caption still has "Bottom Panel" instead of "Output Panel" currently.

File: plugin/__init__.py [hunk 3]
```diff
@@ -12,6 +12,7 @@
 from .core.registry import LspTextCommand
 from .core.registry import LspWindowCommand
 from .core.sessions import AbstractPlugin
+from .core.sessions import AbstractPluginV2
```
Comment (predragnikolic) [inline plugin/__init__.py:15]: accidentally committed? :)

File: plugin/references.py [hunk 4]
```diff
@@ -113,17 +130,27 @@ def _handle_response(
...
+        if output_mode is None:
+            show_in_quick_panel = userprefs().show_references_in_quick_panel
+            if modifier_keys.get('shift'):
+                show_in_quick_panel = not show_in_quick_panel
+        else:
+            show_in_quick_panel = output_mode == 'quick_panel'
+        if show_in_quick_panel:
+            if modifier_keys.get('primary'):
+                side_by_side = True
+            self._show_references_in_quick_panel(word, session, response, side_by_side, force_group, group, position)
```
Comment (predragnikolic) [inline plugin/references.py:151]: this part was a bit hard to look a and understand, but I guess it is necessary if we want to support `primary` and `shift`

File: plugin/__init__.py (hunk 3 above)
Comment (rchl) [reply plugin/__init__.py]: Yeah, hard at work on the new API ;)

**Review by jwortmann — APPROVED** [review]

**Pull request merged**